File: src/justitone/jidi/JidiSequence.java
package justitone.jidi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import org.apache.commons.math3.fraction.BigFraction;

import justitone.Event;
import justitone.Sequence;
import justitone.Song;

//java intonation digital interface
public class JidiSequence {
    public List<JidiTrack> tracks;
    private Map<JidiTrack, Periods> used;

    public final int ppm;
    public final int bpm;

    public JidiSequence(Song song, int ppq) {
        bpm = song.bpm;
        ppm = ppq*4;
        
        tracks = new ArrayList<>();
        used = new HashMap<>();

        JidiTrack track = allocateTrack(BigFraction.ZERO, song.sequence.length());

        loadSequence(new State(), BigFraction.ZERO, false, song.sequence, track);
    }

    public JidiTrack allocateTrack(BigFraction start, BigFraction end) {
        Optional<JidiTrack> maybeTrack = tracks.stream()
                                               .filter(t -> used.get(t).canAllocate(start, end))
                                               .findFirst();
        if (maybeTrack.isPresent()) {
            JidiTrack track = maybeTrack.get();
            
            used.get(track).allocate(start, end);
            
            return track;
        }
        else {
            JidiTrack track = new JidiTrack(tracks.size());
            Periods periods = new Periods();
            periods.allocate(start, end);
            tracks.add(track);
            used.put(track, periods);
            return track;
        }
    }
    
    public void loadSequence(State state, BigFraction currentPos, boolean noteOn, Sequence sequence, JidiTrack track) {
        track.events.stream().filter(e -> e instanceof JidiEvent.Instrument).collect(Collectors.toList());
        track.add(new JidiEvent.Instrument(currentPos.multiply(ppm).intValue(), state.instrument));
        
        for (Event e : sequence.contents()) {
            long tick = currentPos.multiply(ppm).longValue();

            if (e instanceof Event.Note) {
                track.add(new JidiEvent.Token(tick, e.tokens));
                
                if (noteOn) {
                    track.add(new JidiEvent.NoteOff(tick));
                }

                float freq = ((Event.Note) e).ratio().multiply(state.freqMultiplier).floatValue() * 440f;

                track.add(new JidiEvent.Pitch(tick, freq));
                track.add(new JidiEvent.NoteOn(tick));
                
                noteOn = true;
            }
            else if (e instanceof Event.Rest) {
                track.add(new JidiEvent.Token(tick, e.tokens));
                
                if (noteOn) {
                    track.add(new JidiEvent.NoteOff(tick));
                }

                noteOn = false;
            }
            else if (e instanceof Event.Hold) {
                track.add(new JidiEvent.Token(tick, e.tokens));
            }
            else if (e instanceof Event.Modulation) {
                state = state.multiplyFreq(((Event.Modulation) e).ratio());
            }
            else if (e instanceof Event.Instrument) {
                if (noteOn) {
                    track.add(new JidiEvent.NoteOff(tick));
                }
                
                Event.Instrument i = (Event.Instrument) e;
                track.add(new JidiEvent.Instrument(tick, i.instrument));
                
                state = state.changeInstrument(((Event.Instrument) e).instrument);
                
                noteOn = false;
            }
            else if (e instanceof Event.SubSequence) {
                Event.SubSequence sub = (Event.SubSequence) e;
                loadSequence(state.multiplyLength(sub.eventLength()).multiplyFreq(sub.ratio()), 
                             currentPos, noteOn, sub.sequence(), track);
            }
            else if (e instanceof Event.Poly) {
                Event.Poly poly = (Event.Poly) e;
                loadPoly(state, currentPos, noteOn, poly.sequences, track);
            }

            currentPos = currentPos.add(e.length().multiply(state.lengthMultiplier));
        }

        long tick = currentPos.multiply(ppm).longValue();
        
        track.add(new JidiEvent.Token(tick, Collections.emptyList()));
        if(noteOn) {
            track.add(new JidiEvent.NoteOff(currentPos.multiply(ppm).intValue()));
        }
    }
    
    public void loadPoly(State state, BigFraction currentPos, boolean noteOn, List<Event.SubSequence> subs, JidiTrack track) {
        Event.SubSequence sub = subs.get(0);
        
        loadSequence(state.multiplyLength(sub.eventLength()).multiplyFreq(sub.ratio()), 
                     currentPos, noteOn, subs.get(0).sequence(), track);
        
        if (subs.size() > 1) {
            for (int i = 1; i < subs.size(); i++) {
                sub = subs.get(i);
                
                BigFraction end = currentPos.add(sub.length().multiply(state.lengthMultiplier));
                
                loadSequence(state.multiplyLength(sub.eventLength()).multiplyFreq(sub.ratio()), 
                             currentPos, false, sub.sequence(), allocateTrack(currentPos, end));
            }
        }
    }
    
    private class State {
        BigFraction lengthMultiplier;
        BigFraction freqMultiplier;
        int instrument;
        
        public State(BigFraction lengthMultiplier, BigFraction freqMultiplier, int instrument) {
            this.lengthMultiplier = lengthMultiplier;
            this.freqMultiplier = freqMultiplier;
            this.instrument = instrument;
        }
        
        public State(State state) {
            this(state.lengthMultiplier, state.freqMultiplier, state.instrument);
        }
        
        public State() {
            this(BigFraction.ONE, BigFraction.ONE, 0);
        }
        
        public State multiplyLength(BigFraction multiplier) {
            State state = new State(this);
            state.lengthMultiplier = lengthMultiplier.multiply(multiplier);
            return state;
        }
        
        public State multiplyFreq(BigFraction multiplier) {
            State state = new State(this);
            state.freqMultiplier = freqMultiplier.multiply(multiplier);
            return state;
        }
        
        public State changeInstrument(int instrument) {
            State state = new State(this);
            state.instrument = instrument;
            return state;
        }
    }
}

File: src/main/kotlin/justitone/jidi/JidiSequence.kt
package justitone.jidi

import justitone.Event
import justitone.Song
import justitone.util.plus
import justitone.util.times
import org.apache.commons.math3.fraction.BigFraction
import java.lang.RuntimeException

//java intonation digital interface
class JidiSequence(song: Song = Song(Event.Group(), 120), ppq: Int = 768) {
    var tracks: MutableList<JidiTrack> = mutableListOf()
    private val used: MutableMap<JidiTrack, Periods> = mutableMapOf()

    val ppm: Int = ppq * 4;
    val bpm: Int = song.bpm;

    init {
        val track = allocateTrack(BigFraction.ZERO, song.event.length())

        loadEvent(State(), song.event, track)
    }

    private fun allocateTrack(start: BigFraction, end: BigFraction): JidiTrack {
        val track = tracks.asSequence()
                .filter { t -> t.periods.canAllocate(start, end) }
                .firstOrNull()
        return if (track != null) {
            track.periods.allocate(start, end)

            track
        } else {
            val newTrack = JidiTrack(tracks.size)
            newTrack.periods.allocate(start, end)
            tracks.add(newTrack)

            newTrack
        }
    }

    private fun loadEvent(state: State, e: Event, track: JidiTrack) : State {
        val tick = state.currentTick(ppm);

//        track.add(JidiEvent.Instrument(state.currentTick(ppm), state.instrument))

        return when (e) {
            is Event.Leaf -> {
                track.add(JidiEvent.Token(tick, e.tokens))

                if (state.noteOn) {
                    track.add(JidiEvent.NoteOff(tick))
                }

                if (e.ratio == BigFraction.ZERO) {
                    state.advancePos(e.length())
                } else {
                    val freq = e.ratio.toFloat() * 440f

                    track.add(JidiEvent.Pitch(tick, freq))
                    track.add(JidiEvent.NoteOn(tick))

                    state.copy(noteOn = true).advancePos(e.length())
                }
            }
            is Event.Group -> {
                if (state.noteOn) {
                    track.add(JidiEvent.NoteOff(tick))
                }

                val newState = e.children.fold(state.copy(noteOn = false)) { state, event ->
                    loadEvent(state, event, track)
                }

                track.add(JidiEvent.Token(newState.currentTick(ppm), emptySet()))
                if (newState.noteOn) {
                    track.add(JidiEvent.NoteOff(newState.currentTick(ppm)))
                }

                newState
            }
            is Event.PolyGroup -> {
                if (state.noteOn) {
                    track.add(JidiEvent.NoteOff(tick))
                }

                loadPoly(state.copy(noteOn = false), e, track)

                state.advancePos(e.length())
            }
            else -> {
                throw RuntimeException()
            }
        }
    }

    private fun loadPoly(state: State, poly: Event.PolyGroup, track: JidiTrack) {
        var event = poly.children[0]

        loadEvent(state, event, track)

        if (poly.children.size > 1) {
            for (i in 1 until poly.children.size) {
                event = poly.children[i]

                val end = state.currentPos.add(event.length())

                loadEvent(state, event, allocateTrack(state.currentPos, end))
            }
        }
    }

    private data class State(val currentPos: BigFraction = BigFraction.ZERO,
                             val noteOn: Boolean = false) {
        fun currentTick(ppm: Int): Long = (currentPos * ppm).toLong()
        fun advancePos(time: BigFraction): State = copy(currentPos = currentPos + time)
    }
}
